source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -0,0 +1,214 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "extensions/transport_sockets/alts/tsi_socket.h"
+
+#include "common/common/assert.h"
+#include "common/common/enum_to_int.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Alts {
+
+TsiSocket::TsiSocket(HandshakerFactory handshaker_factory, HandshakeValidator handshake_validator)
+    : handshaker_factory_(handshaker_factory), handshake_validator_(handshake_validator),
+      raw_buffer_callbacks_(*this) {
+  raw_buffer_socket_.setTransportSocketCallbacks(raw_buffer_callbacks_);
+}
+
+TsiSocket::~TsiSocket() { ASSERT(!handshaker_); }
+
+void TsiSocket::setTransportSocketCallbacks(Envoy::Network::TransportSocketCallbacks& callbacks) {
+  callbacks_ = &callbacks;
+
+  handshaker_ = handshaker_factory_(callbacks.connection().dispatcher());
```

Comment:
How hard would it be to move handshaker creation to somewhere later, such as doHandshake()? I'm asking this because we want to pass in source/peer address as arguments to this factory function for our extension. But source address isn't known till we call connect(). 

If yes, I'd appreciate if you can add source/peer address into the arguments too. If it will add a lot more complication to this PR, you or I can do these change as follow-up.

Reply:
Yeah it is definitely possible. Anyway I will break this PR to small PRs and keep this PR as draft big picture.